File: POTD/9AUG2026.cpp
/*
Problem: 1140. Stone Game II
Difficulty: Medium
Topic: Dynamic Programming, Recursion, Game Theory

Approach:
- Use DP with three states:
  person -> whose turn it is
  i      -> current pile index
  M      -> maximum number of piles allowed based on previous moves
- Alice tries to maximize her total stones.
- Bob tries to minimize Alice's total stones.
- For every turn, try taking X piles where 1 <= X <= 2*M.
- After taking X piles, update M = max(M, X).
- Memoization avoids recalculating the same state.

Time Complexity: O(n³)
Space Complexity: O(n²)
*/

class Solution {
public:
    int vec[2][101][101];

    int solve(int person, int i, int M, vector<int>& piles) {
        int n = piles.size();

        if (i >= n)
            return 0;

        if (vec[person][i][M] != -1)
            return vec[person][i][M];

        int result = (person == 1) ? -1 : INT_MAX;
        int stones = 0;

        for (int x = 1; x <= min(2 * M, n - i); x++) {

            if (person == 1) {
                stones += piles[i + x - 1];

                result = max(
                    result,
                    stones + solve(0, i + x, max(M, x), piles)
                );
            }
            else {
                result = min(
                    result,
                    solve(1, i + x, max(M, x), piles)
                );
            }
        }

        return vec[person][i][M] = result;
    }

    int stoneGameII(vector<int>& piles) {
        memset(vec, -1, sizeof(vec));

        return solve(1, 0, 1, piles);
    }
};

File: POTD/9AUG2026.py
"""
Problem: 1140. Stone Game II
Difficulty: Medium
Topic: Dynamic Programming, Recursion, Game Theory

Approach:
- Use DP with three states:
  person -> whose turn it is
  i      -> current pile index
  M      -> maximum number of piles allowed
- Alice maximizes her total stones.
- Bob minimizes Alice's total stones.
- On each turn, try taking X piles where 1 <= X <= 2*M.
- Update M = max(M, X).
- Memoization avoids recalculating states.

Time Complexity: O(n³)
Space Complexity: O(n²)
"""

class Solution:
    def solve(self, person, i, M, piles, dp):
        n = len(piles)

        if i >= n:
            return 0

        if dp[person][i][M] != -1:
            return dp[person][i][M]

        if person == 1:
            result = -1
        else:
            result = float('inf')

        stones = 0

        for x in range(1, min(2 * M, n - i) + 1):

            if person == 1:
                stones += piles[i + x - 1]

                result = max(
                    result,
                    stones + self.solve(
                        0, i + x, max(M, x), piles, dp
                    )
                )

            else:
                result = min(
                    result,
                    self.solve(
                        1, i + x, max(M, x), piles, dp
                    )
                )

        dp[person][i][M] = result
        return result

    def stoneGameII(self, piles: List[int]) -> int:
        n = len(piles)

        dp = [[[-1] * (n + 1) for _ in range(n)]
              for _ in range(2)]

        return self.solve(1, 0, 1, piles, dp)
